tools/svn2cl/svn2cl.1
.\" Copyright (C) 2005, 2006, 2007, 2008, 2009, 2010 Arthur de Jong
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior
.\"    written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
.\" GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
.\" IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\" OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
.\" IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.TH "svn2cl" "1" "Oct 2010" "Version 0.13" "User Commands"
.nh
.SH "NAME"
svn2cl - Create a ChangeLog from a subversion log.
.SH "SYNOPSIS"
.B svn2cl
.RI [ OPTION ]...
.RI [ PATH ]...

.SH "DESCRIPTION"
\fBsvn2cl\fP generates a classic GNU-style ChangeLog from the log messages in a subversion repository.
It acts as a wrapper around the '\fBsvn log\fP' command, parsing the XML output with an XSLT stylesheet.
Alternatively it can generate HTML output intended for use with the \fBsvn2html.css\fP style sheet that comes with \fBsvn2cl\fP.

In addition to its own options, it accepts and passes along most svn log options; see '\fBsvn help log\fP' for a list of those and their documentation.
If PATH is not given, \fBsvn2cl\fP will run svn log on the current directory, so it should do the right thing when run from the top of a subversion checkout of the project.

.TP
\fB\-\-strip\-prefix\fR=\fINAME\fR
Strip \fINAME\fR from the first part of all file names (e.g. project/trunk).
By defaults the current path inside the repository is stripped.

.TP
\fB\-\-linelen\fR=\fINUM\fR
Wrap lines at \fINUM\fR characters.
By default, lines are wrapped at 75 characters.
This option is ignored when the \-\-html option is specified.

.TP
\fB\-\-group\-by\-day\fR
This option causes all commit messages to be grouped by day, as long as all the changes are by the same author.
By default each message is listed separately with its own timestamp.

.TP
\fB\-\-separate\-daylogs\fR
Include a blank line between log entries when they are grouped by day.
Only useful with the \fB\-\-group\-by\-day\fR option.
This option is ignored when the \-\-html option is specified, edit \fBsvn2html.css\fP to tune the layout.

.TP
\fB\-i\fR, \fB\-\-include\-rev\fR
Include the subversion revision number in the ChangeLog.
If the \-\-html option is specified this is implied, edit \fBsvn2html.css\fP to turn it off.

.TP
\fB\-a\fR, \fB\-\-include\-actions\fR
Add action labels [ADD], [DEL] and [CPY] tags to files to show the operation
that is performned on the files.

.TP
\fB\-\-break\-before\-msg\fR[=\fINUM\fR]
Add a line break (or multiple line breadks) between the paths and the log message.
If the \-\-html option is specified this option is ignored, edit \fBsvn2html.css\fP to set spacing.

.TP
\fB\-\-reparagraph\fR
Newlines inside paragraphs are ignored and paragraphs are re-wrapped.
Paragraphs are separated by empty lines.

.TP
\fB\-\-title\fR=\fINAME\fR
If outputting an HTML file the \fINAME\fR is used as the title.
This option is ignored for normal text output.

.TP
\fB\-\-revision\-link\fR=\fINAME\fR
This option is used to generate links from the revision numbers in the generated HTML file.
If \fINAME\fR contains two hashes '\fB##\fP' that part will be replaced by the revision number, otherwise the revision number will be appended to \fINAME\fR.
Only the first occurrence of '\fB##\fP' will be replaced.
.br
This option is ignored for normal text output.

.TP
\fB\-\-ticket\-link\fR=\fINAME\fR
This option is used to generate links from references to tickets within the commit message in the generated HTML file.
Anything that looks like a ticket reference (e.g. '\fB#foo\fP') will be replaced with a link to NAME with the ticket part ('\fBfoo\fP' in the example) appended.
.br
This option is ignored for normal text output.

.TP
\fB\-\-ignore\-message\-starting\fR=\fISTRING\fR
Any log messages that start with the specified \fISTRING\fR are ignored and will not show up in the output.
The STRING comparison is case sensitive.

.TP
\fB\-o\fR, \fB\-f\fR, \fB\-\-output\fR=\fIFILE\fR, \fB\-\-file\fR=\fIFILE\fR
Output ChangeLog to \fIFILE\fR instead of the default \fBChangeLog\fP (or \fBChangeLog.html\fP for HTML output).

.TP
\fB\-\-stdout\fR
Output ChangeLog to stdout instead of to a file.

.TP
\fB\-\-authors\fR=\fIFILE\fR
The \fIFILE\fR is used to map author names as exported by svn to full names.
See the section on AUTHORS FILES below for more information.
.br
Note that the \-\-authors option can have poor performance in some cases.

.TP
\fB\-\-html\fR
Generate an HTML file containing a ChangeLog instead of the default text ChangeLog file.
This feature is still experimental.
The \fBsvn2html.css\fP file that is included in the distribution can be used to format the HTML.

.TP
\fB\-h\fR, \fB\-\-help\fR
Show short summary of options.

.TP
\fB\-V\fR, \fB\-\-version\fR
Show version of program.

.PP
Optional PATH arguments and the following options are passed directly to the '\fBsvn \-\-xml log\fP' command:
\fB\-r\fR, \fB\-\-revision\fR, \fB\-\-targets\fR, \fB\-\-stop\-on\-copy\fR, \fB\-\-username\fR, \fB\-\-password\fR, \fB\-\-no\-auth\-cache\fR, \fB\-\-non\-interactive\fR, \fB\-\-config\-dir\fR and \fB\-\-limit\fR.
.br
See '\fBsvn help log\fP' for more information on these options.

.SH "AUTHORS FILES"

An authors file is basically an XML file that should look something like:
.ft B
    <?xml version="1.0" encoding="utf\-8"?>
    <authors>
      <author uid="arthur">
        Arthur de Jong &lt;arthur@arthurdejong.org&gt;
      </author>
    </authors>
.ft R
.br
svn2cl searches for <author> tags with the specified uid attribute.
The content from the author tag is substituted for the author name.

The <author> tag may also contain XML child tags which can be useful for embedding HTML into the result.
For this to work namespace information should be included for HTML output:
.ft B
    <?xml version="1.0" encoding="utf\-8"?>
    <authors xmlns:html="http://www.w3.org/1999/xhtml">
      <author uid="arthur">
        Arthur de Jong &lt;<html:b>arthur@arthurdejong.org</html:b>&gt;
      </author>
    </authors>
.ft R
.br
The svn2cl distribution contains a sample authors.xml file.

As a bonus a plain text authors file that looks like the following is automatically converted to the XML representation:
.ft B
    arthur:Arthur de Jong <arthur@arthurdejong.org>
.ft R

.SH "NOTES"

Note that the default revison range for \fBsvn log\fP is BASE:1.
This means that \fBsvn2cl\fP will not always include the most recent revisons
of the repository. Either run '\fBsvn update\fP' before running \fBsvn2cl\fP
or pass '\fB\-r HEAD:1\fP'.

.SH "EXAMPLES"

Run svn log recursively on the current directory and generate a text ChangeLog file in the current directory for the entire log history:
.ft B
    svn2cl
.ft R

Output the changes from the last week, while including revision numbers in the ChangeLog:
.ft B
    svn2cl \-i \-r "HEAD:{`date \-d '7 days ago' '+%F %T'`}"
.ft R
.br
(actually subversion will include the first revision before the specified period)

Generate an HTML changelog for svn://svn.example.com/project/trunk, removing "project/trunk" from the file names in the changelog.
Group all commits per day into one ChangeLog entry and only include changes from 2005:
.ft B
    svn2cl \-\-html \-\-strip\-prefix=project/trunk \-\-group\-by\-day \e
           \-r '{2006\-01\-01}:{2005\-01\-01}' \e
           svn://svn.example.com/project/trunk
.ft R

.SH "SEE ALSO"
.BR '\fBsvn\ help\ log\fP'

.SH "COPYRIGHT"
Copyright \(co 2005, 2006, 2007, 2008, 2009, 2010 Arthur de Jong <arthur@arthurdejong.org>.
.br
This is free software; see the license for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


tools/svn2cl/svn2cl.sh
#!/bin/sh

# svn2cl.sh - front end shell script for svn2cl.xsl, calls xsltproc
#             with the correct parameters
#
# Copyright (C) 2005, 2006, 2007, 2008, 2009, 2010 Arthur de Jong.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materials provided with the
#    distribution.
# 3. The name of the author may not be used to endorse or promote
#    products derived from this software without specific prior
#    written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
# GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
# IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
# IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

# exit on any failures
set -e
# report unset variables
set -u

# svn2cl version
VERSION="0.13"

# set default parameters
PWD=`pwd`
STRIPPREFIX="AUTOMATICALLY-DETERMINED"
LINELEN=75
GROUPBYDAY="no"
INCLUDEREV="no"
BREAKBEFOREMSG="no"
REPARAGRAPH="no"
SEPARATEDAYLOGS="no"
ACTIONS="no"
CHANGELOG=""
OUTSTYLE="cl"
SVNLOGCMD="svn --verbose --xml log"
SVNINFOCMD="svn --non-interactive info"
AUTHORSFILE=""
IGNORE_MESSAGE_STARTING=""
TITLE="ChangeLog"
REVISION_LINK="#r"
TICKET_LINK=""
TMPFILES=""
AWK="awk"

# do command line checking
prog=`basename "$0"`
while [ $# -gt 0 ]
do
  case "$1" in
    --strip-prefix)
      STRIPPREFIX="$2"
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --strip-prefix=*)
      STRIPPREFIX=`echo "$1" | sed 's/^--[a-z-]*=//'`
      shift
      ;;
    --linelen)
      LINELEN="$2";
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --linelen=*)
      LINELEN=`echo "$1" | sed 's/^--[a-z-]*=//'`
      shift
      ;;
    --group-by-day)
      GROUPBYDAY="yes";
      shift
      ;;
    --separate-daylogs)
      SEPARATEDAYLOGS="yes"
      shift
      ;;
    -i|--include-rev)
      INCLUDEREV="yes";
      shift
      ;;
    -a|--include-actions)
      ACTIONS="yes"
      shift
      ;;
    --break-before-msg|--breaks-before-msg)
      # FIXME: if next argument is numeric use that as a parameter
      BREAKBEFOREMSG="yes"
      shift
      ;;
    --break-before-msg=*|--breaks-before-msg=*)
      BREAKBEFOREMSG=`echo "$1" | sed 's/^--[a-z-]*=//'`
      shift
      ;;
    --reparagraph)
      REPARAGRAPH="yes"
      shift
      ;;
    --title)
      TITLE="$2"
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --title=*)
      TITLE=`echo "$1" | sed 's/^--[a-z-]*=//'`
      shift
      ;;
    --revision-link)
      REVISION_LINK="$2"
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --revision-link=*)
      REVISION_LINK=`echo "$1" | sed 's/^--[a-z-]*=//'`
      shift
      ;;
    --ticket-link)
      TICKET_LINKK="$2"
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --ticket-link=*)
      TICKET_LINK=`echo "$1" | sed 's/^--[a-z-]*=//'`
      shift
      ;;
    --ignore-message-starting)
      IGNORE_MESSAGE_STARTING="$2"
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --ignore-message-starting=*)
      IGNORE_MESSAGE_STARTING=`echo "$1" | sed 's/^--[a-z-]*=//'`
      shift
      ;;
    -f|--file|-o|--output)
      CHANGELOG="$2"
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --file=*|--output=*)
      CHANGELOG=`echo "$1" | sed 's/^--[a-z-]*=//'`
      shift
      ;;
    --stdout)
      CHANGELOG="-"
      shift
      ;;
    --authors)
      AUTHORSFILE="$2"
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --authors=*)
      AUTHORSFILE=`echo "$1" | sed 's/^--[a-z-]*=//'`
      shift
      ;;
    --html)
      OUTSTYLE="html"
      shift
      ;;
    -r|--revision|-c|--change|--targets|-l|--limit)
      # add these as extra options to the log command (with argument)
      arg=`echo "$2" | sed "s/'/'\"'\"'/g"`
      SVNLOGCMD="$SVNLOGCMD $1 '$arg'"
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --revision=*|--change=*|--targets=*|--limit=*)
      # these are single argument versions of the above (with argument)
      arg=`echo "$1" | sed "s/'/'\"'\"'/g"`
      SVNLOGCMD="$SVNLOGCMD '$arg'"
      shift
      ;;
    --username|--password|--config-dir|--config-option)
      # add these as extra options to the log and info commands (with argument)
      arg=`echo "$2" | sed "s/'/'\"'\"'/g"`
      SVNLOGCMD="$SVNLOGCMD $1 '$arg'"
      SVNINFOCMD="$SVNINFOCMD $1 '$arg'"
      shift 2 || { echo "$prog: option requires an argument -- $1";exit 1; }
      ;;
    --username=*|--password=*|--config-dir=*|--config-option=*)
      # these are single argument versions of the above (with argument)
      arg=`echo "$1" | sed "s/'/'\"'\"'/g"`
      SVNLOGCMD="$SVNLOGCMD '$arg'"
      SVNINFOCMD="$SVNINFOCMD '$arg'"
      shift
      ;;
    -g|--use-merge-history|--stop-on-copy)
      # add these as simple options to the log command
      SVNLOGCMD="$SVNLOGCMD $1"
      shift
      ;;
    --no-auth-cache|--non-interactive|--trust-server-cert)
      # add these as simple options to both the log and info commands
      SVNLOGCMD="$SVNLOGCMD $1"
      SVNINFOCMD="$SVNINFOCMD $1"
      shift
      ;;
    -V|--version)
      echo "$prog $VERSION";
      echo "Written by Arthur de Jong."
      echo ""
      echo "Copyright (C) 2005, 2006, 2007, 2008, 2009, 2010 Arthur de Jong."
      echo "This is free software; see the source for copying conditions.  There is NO"
      echo "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE."
      exit 0
      ;;
    -h|--help)
      echo "Usage: $prog [OPTION]... [PATH]..."
      echo "Generate a ChangeLog from a subversion repository."
      echo ""
      echo "  --strip-prefix=NAME  prefix to strip from all entries, defaults path"
      echo "                       inside the repository"
      echo "  --linelen=NUM        maximum length of an output line"
      echo "  --group-by-day       group changelog entries by day"
      echo "  --separate-daylogs   put a blank line between grouped by day entries"
      echo "  -i, --include-rev    include revision numbers"
      echo "  -a, --include-actions     add [ADD], [DEL] and [CPY] tags to files"
      echo "  --break-before-msg[=NUM]  add a line break (or multiple breaks)"
      echo "                       between the paths and the log message"
      echo "  --reparagraph        rewrap lines inside a paragraph"
      echo "  --title=NAME         title used in html file"
      echo "  --revision-link=NAME link revision numbers in html output"
      echo "  --ticket-link=NAME   change #foo strings to links"
      echo "  --ignore-message-starting=STRING"
      echo "                       ignore messages starting with the string"
      echo "  -o, --output=FILE    output to FILE instead of ChangeLog"
      echo "  -f, --file=FILE      alias for -o, --output"
      echo "  --stdout             output to stdout instead of ChangeLog"
      echo "  --authors=FILE       file to read for authors"
      echo "  --html               output as html instead of plain text"
      echo "  -h, --help           display this help and exit"
      echo "  -V, --version        output version information and exit"
      echo ""
      echo "PATH arguments and the following options are passed to the svn log"
      echo "command: -r, --revision, -g, --use-merge-history, -c, --change,"
      echo "--targets, --stop-on-copy, -l, --username, --password, --no-auth-cache,"
      echo "--non-interactive, --trust-server-cert, --config-dir and --config-option"
      echo "(see 'svn help log' for more information)."
      exit 0
      ;;
    -*)
      echo "$prog: invalid option -- $1"
      echo "Try '$prog --help' for more information."
      exit 1
      ;;
    *)
      arg=`echo "$1" | sed "s/'/'\"'\"'/g"`
      SVNLOGCMD="$SVNLOGCMD '$arg'"
      SVNINFOCMD="$SVNINFOCMD '$arg'"
      shift
      ;;
  esac
done

# find the directory that this script resides in
prog="$0"
while [ -h "$prog" ]
do
  dir=`dirname "$prog"`
  prog=`ls -ld "$prog" | sed "s/^.*-> \(.*\)/\1/;/^[^/]/s,^,$dir/,"`
done
dir=`dirname "$prog"`
dir=`cd "$dir" && pwd`
XSL="$dir/svn2${OUTSTYLE}.xsl"

# check if the authors file is formatted as a legacy
# colon separated file
if [ -n "$AUTHORSFILE" ] && \
    egrep '^(#.*|[a-zA-Z0-9].*:)' "$AUTHORSFILE" > /dev/null 2>/dev/null
then
  # create a temporary file
  tmpfile=`mktemp -t svn2cl.XXXXXX 2> /dev/null || tempfile -s .svn2cl 2> /dev/null || echo "$AUTHORSFILE.$$.xml"`
  arg=`echo "$tmpfile" | sed "s/'/'\"'\"'/g"`
  TMPFILES="$TMPFILES '$arg'"
  # generate an authors.xml file on the fly
  echo '<authors>' > "$tmpfile"
  sed -n 's/&/\&amp;/g;s/</\&lt;/g;s/>/\&gt;/g;s|^\([a-zA-Z0-9][^:]*\):\(.*\)$| <author uid="\1">\2</author>|p' \
      < "$AUTHORSFILE"  >> "$tmpfile"
  echo '</authors>' >> "$tmpfile"
  AUTHORSFILE="$tmpfile"
fi

# find the absolute path of the authors file
# (otherwise xsltproc will find the file relative to svn2cl.xsl)
pwd=`pwd`
AUTHORSFILE=`echo "$AUTHORSFILE" | sed "/^[^/]/s|^|$pwd/|"`

# if no filename was specified, make one up
if [ -z "$CHANGELOG" ]
then
  CHANGELOG="ChangeLog"
  if [ "$OUTSTYLE" != "cl" ]
  then
    CHANGELOG="$CHANGELOG.$OUTSTYLE"
  fi
fi

# try to determin a prefix to strip from all paths
if [ "$STRIPPREFIX" = "AUTOMATICALLY-DETERMINED" ]
then
  STRIPPREFIX=`LANG=C eval "$SVNINFOCMD" | $AWK '/^URL:/{url=$2} /^Repository Root:/{root=$3} END{if(root){print substr(url,length(root)+2)}else{n=split(url,u,"/");print u[n]}}'`
  STRIPPREFIX=`echo "$STRIPPREFIX" | sed 's/%20/ /g'`
fi

# redirect stdout to the changelog file if needed
if [ "x$CHANGELOG" != "x-" ]
then
  exec > "$CHANGELOG"
fi

# actually run the command we need
eval "$SVNLOGCMD" | \
  xsltproc --stringparam strip-prefix "$STRIPPREFIX" \
           --stringparam linelen "$LINELEN" \
           --stringparam groupbyday "$GROUPBYDAY" \
           --stringparam separate-daylogs "$SEPARATEDAYLOGS" \
           --stringparam include-rev "$INCLUDEREV" \
           --stringparam include-actions "$ACTIONS" \
           --stringparam breakbeforemsg "$BREAKBEFOREMSG" \
           --stringparam reparagraph "$REPARAGRAPH" \
           --stringparam authorsfile "$AUTHORSFILE" \
           --stringparam title "$TITLE" \
           --stringparam revision-link "$REVISION_LINK" \
           --stringparam ticket-link "$TICKET_LINK" \
           --stringparam ignore-message-starting "$IGNORE_MESSAGE_STARTING" \
           --nowrite \
           --nomkdir \
           --nonet \
           "$XSL" -

# clean up temporary files
[ -n "$TMPFILES" ] && eval "rm -f $TMPFILES"

# we're done (the previous command could return false)
exit 0
